src/compiler/crystal/types.cr

```diff
@@ -719,6 +719,26 @@ module Crystal
       (yield self) ? true : false
     end
 
+    def always_falsey?
+      remove_indirection.is_a?(NilType)
+    end
+
+    def always_truthy?
+      type = remove_indirection
+      case type
+      when NilType
+        false
+      when BoolType
+        false
+      when PointerInstanceType
+        false
+      when UnionType
+        type.union_types.all? &.always_truthy?
+      else
+        true
+      end
+    end
+
```

Comment:
I believe normal codegen would benefit from this part and it can be extracted into its own PR.

Reply:
Definitely! And this logic/optimization could be applied to While as well but it's currently missing from interpreted and compiled more.

Reply:
I think when I did this, I wanted to get something else working because I didn't have `is_a?` and `responds_to?` working. But it's still a good change :-)